cmd/snap/cmd_routine_portal_info.go

```diff
@@ -0,0 +1,145 @@
+	} `positional-args:"true" required:"true"`
+}
+
+var shortRoutinePortalInfoHelp = i18n.G("Return information about a process")
+var longRoutinePortalInfoHelp = i18n.G(`
+The portal-info command returns information about a process in keyfile format.
+
+This command is used by the xdg-desktop-portal service to retrieve
+information about snap confined processes.
+`)
+
+func init() {
+	addRoutineCommand("portal-info", shortRoutinePortalInfoHelp, longRoutinePortalInfoHelp, func() flags.Commander {
+		return &cmdRoutinePortalInfo{}
+	}, nil, []argDesc{{
+		// TRANSLATORS: This needs to begin with < and end with >
+		name: i18n.G("<process ID>"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		desc: i18n.G("Process ID of confined app"),
+	}})
+}
+
+const portalInfoTemplate = `[Snap Info]
+InstanceName={{.Snap.Name}}
+{{- if .App}}
+AppName={{.App.Name}}
+{{- end}}
+{{- if .DesktopFile}}
+DesktopFile={{.DesktopFile}}
+{{- end}}
+HasNetwork={{.HasNetwork}}
+`
+
+var (
+	cgroupSnapNameFromPid   = cgroup.SnapNameFromPid
+	apparmorSnapNameFromPid = apparmor.SnapNameFromPid
+)
+
+func (x *cmdRoutinePortalInfo) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	snapName, err := cgroupSnapNameFromPid(x.PortalInfoOptions.Pid)
+	if err != nil {
+		return err
+	}
+	snap, _, err := x.client.Snap(snapName)
+	if err != nil {
+		return fmt.Errorf("cannot retrieve info for snap %q: %v", snapName, err)
+	}
+
+	// Try to identify the application name from AppArmor
+	var app *client.AppInfo
+	if snapName, appName, _, err := apparmorSnapNameFromPid(x.PortalInfoOptions.Pid); err == nil && snapName == snap.Name && appName != "" {
+		for i := range snap.Apps {
+			if snap.Apps[i].Name == appName {
+				app = &snap.Apps[i]
+				break
+			}
+		}
+	}
+	// As a fallback, pick an app with a desktop file, favouring
+	// the app named identically to the snap.
+	if app == nil {
+		for i := range snap.Apps {
+			if snap.Apps[i].DesktopFile != "" && (app == nil || snap.Apps[i].Name == snap.Name) {
+				app = &snap.Apps[i]
+			}
+		}
+	}
+
+	var desktopFile string
+	if app != nil {
+		desktopFile = filepath.Base(app.DesktopFile)
+	}
+
+	// Determine whether the snap has access to the network
+	connections, err := x.client.Connections(&client.ConnectionOptions{
+		Snap:      snap.Name,
+		Interface: "network",
+	})
+	if err != nil {
+		return fmt.Errorf("cannot get connections for snap %q: %v", snap.Name, err)
+	}
+	var hasNetwork bool
+	for _, conn := range connections.Established {
+		if conn.Plug.Snap == snap.Name && conn.Interface == "network" {
+			hasNetwork = true
+			break
+		}
+	}
+
+	t := template.Must(template.New("portal-info").Parse(portalInfoTemplate))
+	data := struct {
+		Snap        *client.Snap
+		App         *client.AppInfo
+		DesktopFile string
+		HasNetwork  bool
+	}{
+		Snap:        snap,
+		App:         app,
+		DesktopFile: desktopFile,
+		HasNetwork:  hasNetwork,
+	}
+	return t.Execute(Stdout, data)
```

Comment:
mmh, mabye use a "cannot render output template" kind of error here, instead of the raw template one